### PR TITLE
Add initial attempt for a BTK build on conda-forge

### PR DESCRIPTION
**Pull request by melund**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

File: recipes/btk/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  home: https://github.com/Biomechanical-ToolKit/BTKCore
+  # Remember to specify the license variants for BSD, Apache, GPL, and LGPL.
+  # Prefer the short version, e.g: GPL-2.0 instead of GNU General Public License version 2.0
+  # See https://opensource.org/licenses/alphabetical
+  license: BSD-3-Clause
+  # The license_family, i.e. "BSD" if license is "BSD-3-Clause". (optional)
+  license_family: BSD
+  # It is strongly encouraged to include a license file in the package,
+  # (even if the license doesn't require it) using the license_file entry.
+  # See https://conda.pydata.org/docs/building/meta-yaml.html#license-file
```
Comment (scopatz) [inline recipes/btk/meta.yaml]: Please remove these comments.

File: recipes/btk/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  license_file: LICENSE.txt
+  summary: 'Simple, fast, extensible JSON encoder/decoder for Python'
```
Comment (scopatz) [inline recipes/btk/meta.yaml]: Also, should probably update this :wink:

**Comment by melund** [comment]
Thanks lot @scopatz. It is not quite ready yet. So I don't want to waste your time. I could probably use your help once I have fiddle with this a little more.

**Comment by scopatz** [comment]
Sorry for being too quick here :wink:

**Comment by melund** [comment]
@scopatz. I think I could use you help now. It fails when linking. `LINK : fatal error LNK1104: cannot open file 'python37.lib'` I don't know what is causing this. I don't get the error when I build locally. 

This is from appveyor, but the same error comes on azure.  Which is used these build the windows packages? 

[link](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/29452213#L3079)
```
LINK: command "C:\PROGRA~2\MI0E91~1.0\VC\bin\amd64\link.exe /nologo @CMakeFiles\_btk.dir\objects1.rsp /out:..\..\bin\_btk.pyd /implib:..\..\bin\_btk.lib /pdb:%SRC_DIR%\build\bin\_btk.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO C:\Python38-x64\libs\python38.lib ..\..\bin\BTKIO.lib ..\..\bin\BTKBasicFilters.lib ..\..\bin\BTKCommon.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:..\..\bin\_btk.pyd.manifest" failed (exit code 1104) with the following output:
btkSwigPYTHON_wrap.cxx.obj : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance
LINK : fatal error LNK1104: cannot open file 'python37.lib'
NMAKE : fatal error U1077: 'C:\bld\btk_1576013792819\_build_env\Library\bin\cmake.exe' : return code '0xffffffff'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
```

File: recipes/btk/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  build:
+    - cmake
+    - jom
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - python
+    - swig
+    - numpy >=1.14
+    - setuptools
```
Comment (scopatz) [inline recipes/btk/meta.yaml]: maybe try removing this dep or replacing it withg pip.  I think you get a slightlu different Python depending on which package you depend on here

**Comment by melund** [comment]
@scopatz The recipe is ready and it did build until today. 

Everything build correctly yesterday ([see log](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/29468511)) But today it seems like it never starts the build script (`bld.bat`). 

I don't know what caused this. I did rebased on master, but rolling back didn't seem to fix it. So maybe something else changed on appveyor. Comparing logs I can see that conda and conda build changed versions since yesterday. 

Working (yesterday): conda: 4.6.14  conda-build:  3.18.11 
Not working (today): conda 4.7.10, conda-build: 3.16.3

It also build for for me locally where conda/conda-build are both the newest version (4.7.12 and 3.18.11).

**Comment by scopatz** [comment]
hmmm weird.  Also, conda v4.8 was just released.  CC @msarahan @jjhelmus @soapy1

**Comment by melund** [comment]
I will  try to tigger a new build. Maybe it picks up conda 4.8 today.

**Comment by melund** [comment]
@scopatz The azure pipeline for windows builds successfully again. It is back to using conda build 3.18.11.   Appveyor still never starts the `bld.bat` script. 

Is it still appveyor that creates the Windows packages or can I azure-pipeline for this now?

**Comment by melund** [comment]
@scopatz I force the recipe to use conda-build 3.18, and it now building again on AppVeyor. 
The package is ready now. 

Azure is failing now, with something unrelated. 
```
Error retrieving packages from source 'https://chocolatey.org/api/v2/':
 Could not connect to the feed specified at 'https://chocolatey.org/api/v2/'. Please verify that the package source (located in the Package Manager Settings) is valid and ensure your network connectivity.
```

**Comment by melund** [comment]
Now azure pipeline is back. @scopatz can you merge this if you think the  recipe  looks OK?

File: recipes/btk/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  build:
+    - cmake
+    - jom  # [win]
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - python
+    - swig
+    - numpy >=1.14
+    - vc  14  # [win and py >= 35]
```
Comment (isuruf) [inline recipes/btk/meta.yaml]: Remove this

File: recipes/btk/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    - pip
+  run:
+    - python
+    - numpy
+    - vc  14  # [win and py >= 35]
```
Comment (isuruf) [inline recipes/btk/meta.yaml]: Same here

File: recipes/btk/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+requirements:
+  build:
+    - cmake
+    - jom  # [win]
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - python
+    - swig
+    - numpy >=1.14
```
Comment (isuruf) [inline recipes/btk/meta.yaml]: See https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy

File: recipes/btk/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+#  patches:
+#    - 0001-Updates-since-3.0.patch
+#    - 0002-Python-3-support.patch
+
+
+build:
+  number: 0
+  skip: true  # [not win or py < 36]
+  features: 
+    - vc14  # [win and py >= 35]
```
Comment (isuruf) [inline recipes/btk/meta.yaml]: Remove the 2 lines above

File: recipes/btk/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/Biomechanical-ToolKit/BTKCore/archive/v0.3.zip
+  sha256: {{ sha256 }}
+
+# Patches are applied in bld.bat to handle whitespace issues
+#  patches:
+#    - 0001-Updates-since-3.0.patch
```
Comment (isuruf) [inline recipes/btk/meta.yaml]: This is a huge patch. Why is this needed?

File: recipes/btk/meta.yaml (hunk 6 above)
Comment (melund) [reply recipes/btk/meta.yaml]: Done.

File: recipes/btk/meta.yaml (hunk 7 above)
Comment (melund) [reply recipes/btk/meta.yaml]: Thanks...

File: recipes/btk/meta.yaml (hunk 8 above)
Comment (melund) [reply recipes/btk/meta.yaml]: Good question. BTK is a python 2.7 library which have been heavily used in many biomechanics project. It is also unfortunately abandoned. That is very bad since it is holding many projects back. It is also not trivial to build.

The first patch represents all the work by the maintainer since the 0.3 version in 2014 and until it was abandoned end of 2015. Hence the version 0.4.dev of this recipe. I suspect some of this work was also necessary for building with vc14.  

The second patch is all the fixes I could gather to get Python3 support to work.

It is my hope that this would help a lot of people who don't have the option of  building  BTK them self, and a lot of projects which would otherwise be stuck on Python 2.7.

**Comment by melund** [comment]
@isuruf Thanks for your comments. I have updated and all checks still pass. Is it Appveyor or Azure Pipeline that is used on Windows?

**Comment by scopatz** [comment]
It is Azure that is used by default for windows

File: .appveyor.yml [hunk 9]
```diff
@@ -40,7 +40,7 @@ install:
     - cmd: appveyor-retry conda.exe update --yes --quiet conda
 
 
-    - cmd: appveyor-retry conda.exe install --yes --quiet "conda!=4.6.1,<4.7.11a0" conda-forge-pinning conda-forge-ci-setup=2.* networkx=2.3 "conda-build>=3.16"
+    - cmd: appveyor-retry conda.exe install --yes --quiet "conda!=4.6.1,<4.7.11a0" conda-forge-pinning conda-forge-ci-setup=2.* networkx=2.3 "conda-build>=3.18"
```
Comment (scopatz) [inline .appveyor.yml]: Changes outside the `recipes/` dir should be made in a separate PR. Would you mind putting this in, @melund? Thanks!

File: recipes/btk/meta.yaml (hunk 8 above)
Comment (scopatz) [reply recipes/btk/meta.yaml]: We don't usually package dev releases (we have a few issues discussing this). Is there a way to cut another the release?

It terms of the patch itself, if there a way to put it in a gist and then patch it manually in the build script?

**Comment by scopatz** [comment]
Thanks for all of your effort here @melund!

File: .appveyor.yml (hunk 9 above)
Comment (melund) [reply .appveyor.yml]: @isuruf already fixed this in 6b52315314b1a23048d533fb91cb1857378cc355

File: recipes/btk/meta.yaml (hunk 8 above)
Comment (melund) [reply recipes/btk/meta.yaml]: > We don't usually package dev releases (we have a few issues discussing this). Is there a way to cut another the release?
I don't think so. The maintainer hasn't responded to anything the last few years. But I will try to reach out by other means than github. Maybe he could tag the latest commit. That would eliminate the large patch. 

> It terms of the patch itself, if there a way to put it in a gist and then patch it manually in the build script?

That is no problem. I will try that.

**Comment by melund** [comment]
I have put the patches in a gist. But don't merge this yet. If this work I will have to rebase to remove the large patch files from history.

File: recipes/btk/meta.yaml (hunk 8 above)
Comment (isuruf) [reply recipes/btk/meta.yaml]: > It terms of the patch itself, if there a way to put it in a gist and then patch it manually in the build script?

I'm sorry, but what does that gain us?

Comment (melund) [reply recipes/btk/meta.yaml]: Not much, but it does keep the large file out of the staged-recipes git repo.

**Comment by melund** [comment]
@scopatz I have remove the large patch by pointing to the latest commit on github. It is will not be a tagged version but it is still much better that the large patch. 

I know it does not align with the general recommendation on conda-forge; pointing to tagged version... But it is just as secure as a tagged version since we are pointing to specific git hash (not a branch). The sha256 hash in the recipe will also ensure we that we are building the right thing. 

I have reached out to the original author about tagging the latest version. If he responds we can update this in the feedstock. 

The whole point of this recipe is to ensure Python 3 compatibility for a dying project.

**Comment by scopatz** [comment]
This looks good to me @melund! Let me know when you think it is ready to merge!

**Comment by melund** [comment]
Yes. This is ready :) Thanks @scopatz
